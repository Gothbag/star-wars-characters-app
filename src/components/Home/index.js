import React from "react";
import PropTypes from "prop-types";
import {isEmpty, noop} from "lodash";

import SearchBox from "../SearchBox";
import CharacterCard from "../CharacterCard";

class Home extends React.PureComponent {

    static propTypes = {
        character: PropTypes.shape({
        	birthYear: PropTypes.string,
        	gender: PropTypes.string,
            name: PropTypes.string.isRequired
        }),
        loadCharacter: PropTypes.func,
        loadFilms: PropTypes.func,
        loadPlanet: PropTypes.func,
        loadSpecies: PropTypes.func
    }

    static defaultProps = {
    	character: {},
    	loadCharacter: noop,
        loadFilms: noop,
    	loadPlanet: noop,
      	loadSpecies: noop
    }

    componentDidMount() {
    	const { props: { films, loadFilms } } = this;
    	if (isEmpty(films)) {
    		loadFilms();
    	}
    }

    render() {
        const { props: { character, films, loadCharacter, loadPlanet, loadSpecies, planets, species } } = this;
        return (<section className="col-xs-1-1 col-md-1-3 center">
        	<SearchBox handleOnSearch={loadCharacter}/>
        	<CharacterCard character={character} loadPlanet={loadPlanet} loadSpecies={loadSpecies} planets={planets} species={species} films={films}/>
        </section>);
    }
}

export default Home;
