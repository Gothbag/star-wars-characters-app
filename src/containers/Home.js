import { connect } from "react-redux"

import Home from "../components/Home";
import { loadCharacter, loadFilms, loadSingleton } from "../actions/characterData";
import { listConsts } from "../reducers/lists";

const { CHARACTERS, PLANETS, FILMS, SPECIES } = listConsts;

const mapStateToProps = state => {
	const { lists } = state;
	return {
		character: lists[CHARACTERS].focus,
		films: lists[FILMS].elements,
	    planets: lists[PLANETS].elements,
	    species: lists[SPECIES].elements
	}; 
}

const mapDispatchToProps = dispatch => ({
	loadCharacter: searchTerm => dispatch(loadCharacter(searchTerm)),
	loadFilms: () => dispatch(loadFilms()),
    loadPlanet: id => dispatch(loadSingleton(id, "planets", PLANETS)),
    loadSpecies: id => dispatch(loadSingleton(id, "species", SPECIES))
}); 

export default connect(mapStateToProps, mapDispatchToProps)(Home);
