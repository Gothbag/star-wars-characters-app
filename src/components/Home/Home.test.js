import React from 'react';
import Home from './index';

describe("<Home/>", () => {

	const props = {

	};

	beforeEach(() => {
		props.loadFilms = jest.fn();
		props.films = {};
	});

	it('renders without crashing', () => {
		const wrapper = shallow(<Home/>);
		expect(wrapper.find("SearchBox").length).toBe(1);
	});

	it("should call the loadFilms prop if the films prop is empty", () => {
		const wrapper = mount(<Home {...props}/>);
		expect(props.loadFilms.mock.calls.length).toBe(1);
	});

	it("shouldn't call the loadFilms prop if the films prop is empty", () => {
		props.films = { 1: {
			title: "A New Hope"
		} }
		const wrapper = mount(<Home {...props}/>);
		expect(props.loadFilms.mock.calls.length).toBe(0);
	});

});
