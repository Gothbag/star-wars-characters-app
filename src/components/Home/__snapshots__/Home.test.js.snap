// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CharacterCard/> renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Home
    character={Object {}}
    loadCharacter={[Function]}
    loadFilms={[Function]}
    loadPlanet={[Function]}
    loadSpecies={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <SearchBox
          handleOnClick={[Function]}
/>,
        <CharacterCard
          character={Object {}}
          films={Object {}}
          loadPlanet={[Function]}
          loadSpecies={[Function]}
          planets={Object {}}
          species={Object {}}
/>,
      ],
      "className": "col-xs-1-1 col-md-1-3 center",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "handleOnClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "character": Object {},
          "films": Object {},
          "loadPlanet": [Function],
          "loadSpecies": [Function],
          "planets": Object {},
          "species": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <SearchBox
            handleOnClick={[Function]}
/>,
          <CharacterCard
            character={Object {}}
            films={Object {}}
            loadPlanet={[Function]}
            loadSpecies={[Function]}
            planets={Object {}}
            species={Object {}}
/>,
        ],
        "className": "col-xs-1-1 col-md-1-3 center",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "handleOnClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "character": Object {},
            "films": Object {},
            "loadPlanet": [Function],
            "loadSpecies": [Function],
            "planets": Object {},
            "species": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`<Home/> renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Home
    character={Object {}}
    loadCharacter={[Function]}
    loadFilms={[Function]}
    loadPlanet={[Function]}
    loadSpecies={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <SearchBox
          handleOnClick={[Function]}
/>,
        <CharacterCard
          character={Object {}}
          films={Object {}}
          loadPlanet={[Function]}
          loadSpecies={[Function]}
          planets={Object {}}
          species={Object {}}
/>,
      ],
      "className": "col-xs-1-1 col-md-1-3 center",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "handleOnClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "character": Object {},
          "films": Object {},
          "loadPlanet": [Function],
          "loadSpecies": [Function],
          "planets": Object {},
          "species": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <SearchBox
            handleOnClick={[Function]}
/>,
          <CharacterCard
            character={Object {}}
            films={Object {}}
            loadPlanet={[Function]}
            loadSpecies={[Function]}
            planets={Object {}}
            species={Object {}}
/>,
        ],
        "className": "col-xs-1-1 col-md-1-3 center",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "handleOnClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "character": Object {},
            "films": Object {},
            "loadPlanet": [Function],
            "loadSpecies": [Function],
            "planets": Object {},
            "species": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
